lib/wibox/widget/progressbar.lua

```diff
@@ -144,10 +148,31 @@ function progressbar:set_value(value)
     value = value or 0
     local max_value = self._private.max_value
     self._private.value = math.min(max_value, math.max(0, value))
+
+    -- Keep the value. In case the widget is initiated using the declarative
+    -- syntax, the order in which max_value and value is called is unpredictable.
+    if self._private.value ~= value then
+        self._private.value_cache = value
+    else
+        self._private.value_cache = nil
+    end
+
     self:emit_signal("widget::redraw_needed")
     return self
 end
 
+function progressbar:set_max_value(max_value)
+
+    self._private.max_value = max_value
+
+    -- Restore the value that was trimmed due to a race condition
+    if self._private.value_cache and self._private.value_cache <= max_value then
+        self._private.value = self._private.value_cache
+    end
```

Comment:
Hm... How about having `set_value` only do `self._private.value = value` and move the `math.min(max_value, math.max(0, value))` into `:draw()`? I.e. we never trim the value that is saved, but only apply the triming while drawing the widget. That might be simpler to understand than this version.